Fix tz2 signature verification and signing

sbihel opened the pull request.

sbihel commented:
`ring` doesn't seem to offer a way to sign/verify a prehashed message. `ed25519-dalek` does, and the methods we're using atm also to the sha512 (so we do two hashing).

sbihel commented:
Unless we don't need to do the blake2b hash for tz1? @clehner

clehner commented:
blake2b is not needed for tz1. Ed25519 specifies SHA-512: https://datatracker.ietf.org/doc/html/rfc8032#section-5.1

clehner commented:
> blake2b is not needed for tz1

Actually I'm not sure, seeing as the hash step is removed in https://github.com/spruceid/ssi/pull/200/commits/a573b25c90f6604406984922c44b0095ad565213#diff-859afcdde128730c6eb7badc55e61454f63fed6cbb70d7a513f70502420d57cfL1256

Tests should check that the signature in `edpk_jwk_tz_edsig` and `did-pkh/tests/vc-tz1.jsonld` remain valid, as those were produced from a Tezos wallet.

clehner commented:
I think adding the three algorithms is a correct approach given the current API. But if there is a simpler way, that could be good.

If/when LDP in `ssi` is refactored to not require JWK for signing and verification but instead use verification method and proof objects directly, TezosSignature could be updated to use crypto library functions directly instead of going through `ssi::jws`.

RFC 7515 says the algorithm (`alg`) name should either be registered in the IANA JSON Web Signature and Encryption Algorithms registry, or contain a "Collision-Resistant Name" (ES256K-R does not follow this):
- https://datatracker.ietf.org/doc/html/rfc7515#section-4.1.1
- https://datatracker.ietf.org/doc/html/rfc7515#page-6
- https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms

If using the current approach, maybe `EdBlake2b`, `ESBlake2b`, `ESBlake2bK` should be renamed to URIs?

sbihel commented:
The remaining test was failing due to a mismatch of the old EdDSA in the proof and the use of the new EdBlake2b in the JWK. I have allowed those pairs for the time being (only for the verification, and only the provided public key can have an different/old alg).

-----

I agree that specifying a custom `alg` doesn't fit the specs and the purpose of those proofs.

> If/when LDP in ssi is refactored to not require JWK for signing and verification but instead use verification method and proof objects directly, TezosSignature could be updated to use crypto library functions directly instead of going through ssi::jws.

I might be wrong and biased, but that confirms my previous intention of using the trait combination of Signer + Digest for a potential Signature trait. That being said, we still need to handle those VCs in the wild which we have issued already, so I feel like this work wasn't in vain (yes we're creating more mess that way, but at least it's unambiguous).

How do you specify the public key without `publicKeyJwk`?

clehner commented:
> How do you specify the public key without publicKeyJwk?

The other property listed in DID Core is `publicKeyMultibase`: https://www.w3.org/TR/did-core/#dfn-publickeymultibase
(Previously it was [publicKeyBase58](https://w3c-ccg.github.io/security-vocab/#publicKeyBase58) but that is deprecated now.) This property represents a public key encoded as bytes, but it doesn't specify how, which I think leaves it up to the signature suite / verification method specifications (i.e. our choice here). We could use the existing Tezos-style type-prefixed base58-encoded public key, and prepend "z" to make it [multibase](https://datatracker.ietf.org/doc/html/draft-multiformats-multibase-03#page-6)-encoded (e.g. `zedpk...`, `zsppk...`, `zp2pk...`).

> trait combination of Signer + Digest for a potential Signature trait

I would be interested to see more what you mean.

clehner commented on Cargo.toml:
```diff
@@ -35,7 +35,8 @@ base64 = "0.12"
 ed25519-dalek = { version = "1", optional = true }
-rand = { version = "0.7", optional = true }
+rand = { version = "0.8", optional = true }
+rand_old = { package = "rand", version = "0.7" }
```
Is something requiring rand 0.7 still? I only find it mentioned in didkit-wasm.

sbihel commented:
Rebased and squashed.

> We could use the existing Tezos-style type-prefixed base58-encoded public key, and prepend "z" to make it [multibase](https://datatracker.ietf.org/doc/html/draft-multiformats-multibase-03#page-6)-encoded (e.g. `zedpk...`, `zsppk...`, `zp2pk...`).

That would kind of suck because then we would moved the custom logic from the JWK trait (I mean, the signing/verify functions) to a `publicKeyMultibase` trait. Actually now that I write this it feels like using multibase wouldn't save us from having custom logic on how to interpret the key (i.e. you need to know about the signature type to know how to use the key).

> > trait combination of Signer + Digest for a potential Signature trait
> 
> I would be interested to see more what you mean.

Continuing off the above, I thought we could keep the interpretation of the key pure, independent from the signature. The oddities of the signatures would be restrained to a Digest trait which would be responsible on formatting (e.g. hashing) the input data for the Signer. In other words, the Signature trait is responsible for picking the right Digest and the right Signer.

But I don't know if it's feasible, I might be too optimistic on how much we can segregate the hashing from the actual signing.
